app/repositories/SessionProcessFSM.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package repositories
+
+import javax.inject.{Inject, Singleton}
+import core.models.ocelot.{Flow, FlowStage}
+import DefaultSessionRepository._
+import core.models.ocelot.{Label, ScalarLabel}
+
+
+@Singleton
+class SessionProcessFSM @Inject() {
+  // Input
+  // url ,incoming url
+  // priorSp, prior SessionProcess corresponding to the previous url processed. Note. The db record will have the head of the page history updated
+  //          to include incoming url and flowStack Nil, this update is not within priorSp. This is an optimisation, as the most common transition
+  //          is forward in a process containing no Sequences (i.e. the flowStack will always be empty), this transition will result in
+  //          no page history or flowstack updates given the update described has already taken place.
+  // forceForward, true indicates that a url similar to head of prior history (looks like a BACK) should be treated as a forward movement
+
+  // New state == (priorSp + new Pagehistory head) ++ pageHistory and flowStack output updates
+  //
+  // Output
+  // optional backlink to be displayed on page with incoming url
+  // optional page history update
+  // optional flowStack update
+  def apply(url: String, priorSp: SessionProcess, forceForward: Boolean):
```

Comment:
IntelliJ isn't very appreciative of using the method name "apply" here. For some reason it doesn't recognise the method defined in SessionProcessFSM in the class SessionRepository.

Reply:
Its not the apply, its the missing () after the @Inject() it doesnt like, I've added it and we now have our olf favourite on match statements, cyclomatic complexity